test/metabase/models/card_test.clj

```diff
@@ -1173,3 +1175,154 @@
+    ;; :idents are populated on initial insert; update to remove them. (Update does not populate them like insert.)
+    ;; Also remove the generated :entity_id.
+    (t2/update! :model/Card id {:dataset_query (bare-query)
+                                :entity_id     nil})
+    ;; Can't use the one from `with-temp` since it came before the above edit.
+    (let [{eid   :entity_id
+           query :dataset_query} (t2/select-one :model/Card :id id)]
+      (testing "idents should be populated on read"
+        ;; These idents are: kind_EID@stage__index, eg. "aggregation_4QsLuEnriHKkXCWqbPMQ8@0__0"
+        ;; The entity_id is hashed based on created_at, so it's still always different!
+        (is (string? eid))
+        (is (=? (bare-query-exp eid)
+                (:query query)))
+
+        (testing "but not written back to appdb"
+          (let [{raw-query :dataset_query} (t2/select-one "report_card" :id id)]
+            (is (string? raw-query))
+            (is (nil? (str/index-of raw-query "-idents")))))
+
+        (testing "converted to pMBQL"
+          (let [converted   (lib/query (lib.metadata.jvm/application-database-metadata-provider 1) query)
+                agg-idents  (-> query :query :source-query :aggregation-idents)
+                brk-idents  (-> query :query :source-query :breakout-idents)
+                expr-idents (-> query :query :source-query :expression-idents)
+                join-idents (->> query :query :joins (map :ident))
+                expected    {:stages [{:aggregation [[:count {:ident (get agg-idents 0)}]
+                                                     [:sum   {:ident (get agg-idents 1)} some?]]
+                                       :breakout    [[:field      {:ident (get brk-idents 0)} some?]
+                                                     [:expression {:ident (get brk-idents 1)} some?]]
+                                       :expressions [[:+ {:lib/expression-name "yo"
+                                                          :ident               (get expr-idents "yo")}
+                                                      vector? 7]]}
+                                      {:joins [{:alias "a_join"
+                                                :ident (nth join-idents 0)}
+                                               {:alias "another_join"
+                                                :ident (nth join-idents 1)}]}]}
+                exp-legacy  {:query {:source-query {:aggregation-idents agg-idents
+                                                    :breakout-idents    brk-idents
+                                                    :expression-idents  expr-idents}
+                                     :joins [{:alias "a_join"
+                                              :ident (nth join-idents 0)}
+                                             {:alias "another_join"
+                                              :ident (nth join-idents 1)}]}}]
```

Comment:
Why do you use `nth` for joins and `get` for aggs and brks?

Reply:
Because `:aggregation-idents`, `:breakout-idents` and `:expression-idents` are maps from index (or expression name) to ident, so it's `(get agg-idents index)`.

But `join-idents` is (roughly) `(map :ident (:joins query))`, ie. it's a sequence, and `get` doesn't work on it.

Reply:
I think it should be a vector. In that case `get` would work.

Reply:
I actually replaced this with destructuring `[jid1 jid2]` for the join idents.